icicle/include/hash/hash.cuh

```diff
@@ -0,0 +1,257 @@
+#pragma once
+#ifndef HASH_H
+#define HASH_H
+
+#include "gpu-utils/device_context.cuh"
+#include "gpu-utils/error_handler.cuh"
+
+namespace hash {
+  /**
+   * Squeeze states to extract the results.
+   * 1 GPU thread operates on 1 state.
+   *
+   * @param states the states to squeeze
+   * @param number_of_states number of states to squeeze
+   * @param width Width of the state
+   * @param rate Squeeze rate. How many elements to extract from each state
+   * @param offset Squeeze offset. Start squeezing from Oth element of the state
+   * @param out pointer for squeeze results. Can be equal to states to do in-place squeeze
+   *
+   * @tparam S Type of the state element
+   */
+  template <typename S>
+  __global__ void generic_squeeze_states_kernel(
+    const S* states, unsigned int number_of_states, unsigned int width, unsigned int rate, unsigned int offset, S* out)
+  {
+    int idx = (blockIdx.x * blockDim.x) + threadIdx.x;
+    if (idx >= number_of_states) { return; }
+
+    for (int i = 0; i < rate; i++) {
+      if (states == out) {
+        S element = states[idx * width + offset];
+        __syncthreads();
+        out[idx * rate + i] = element;
+      } else {
+        out[idx * rate + i] = states[idx * width + offset];
+      }
+    }
+  }
+
+  template <typename Image>
+  class Hash
+  {
+  public:
+    unsigned int width;
+    unsigned int preimage_max_length;
+
+    Hash() = default;
+    Hash(unsigned int width, unsigned int preimage_max_length) : width(width), preimage_max_length(preimage_max_length)
+    {
+    }
+
+    virtual cudaError_t pad_many(
+      Image* states,
+      unsigned int number_of_states,
+      unsigned int input_block_len,
+      unsigned int rate,
+      const device_context::DeviceContext& ctx) const
+    {
+      return cudaError_t::cudaSuccess;
+    };
+
+    virtual cudaError_t squeeze_states(
+      const Image* states,
+      unsigned int number_of_states,
+      unsigned int rate,
+      unsigned int offset,
+      bool align,
+      Image* output,
+      const device_context::DeviceContext& ctx) const = 0;
+
+    virtual cudaError_t run_permutation_kernel(
```

Comment:
(1) can it be inplace? do you assume that derived class supports both in-place and out-of-place? also for squeeze_states.
(2) what is aligned? to what?

Reply:
1) yes it can. All of our hash functions support in-place permutation and I don't know of a hash that won't support it. For squeeze_states - no, it can't be inplace as you pointed in the kernel comment. Already changed it locally
2) Align refers to domain separation. In poseidon1 for example you need to put a domain tag as the first element of your permutation state. Align boolean is used in a poseidon1 kernel and is ignored for all the other hashes